statistics/handle/handle_test.go

```diff
@@ -1512,6 +1512,49 @@ func (s *testStatsSuite) TestAnalyzeWithDynamicPartitionPruneMode(c *C) {
 	c.Assert(rows[1][6], Equals, "6")
 }
 
+func (s *testStatsSuite) TestPartitionPruneModeSessionVariable(c *C) {
+	defer cleanEnv(c, s.store, s.do)
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk1.MustExec("use test")
+	tk1.MustExec("set @@tidb_partition_prune_mode = '" + string(variable.Dynamic) + "'")
+	tk1.MustExec(`set @@tidb_analyze_version=2`)
+
+	tk2 := testkit.NewTestKit(c, s.store)
+	tk2.MustExec("use test")
+	tk2.MustExec("set @@tidb_partition_prune_mode = '" + string(variable.Static) + "'")
+	tk2.MustExec(`set @@tidb_analyze_version=2`)
+
+	tk1.MustExec(`create table t (a int, key(a)) partition by range(a) 
+					(partition p0 values less than (10),
+					partition p1 values less than (22))`)
+
+	tk1.MustQuery("explain format = 'brief' select * from t").Check(testkit.Rows(
+		"TableReader 10000.00 root partition:all data:TableFullScan",
+		"└─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo",
+	))
+	tk2.MustQuery("explain format = 'brief' select * from t").Check(testkit.Rows(
+		"PartitionUnion 20000.00 root  ",
+		"├─TableReader 10000.00 root  data:TableFullScan",
+		"│ └─TableFullScan 10000.00 cop[tikv] table:t, partition:p0 keep order:false, stats:pseudo",
+		"└─TableReader 10000.00 root  data:TableFullScan",
+		"  └─TableFullScan 10000.00 cop[tikv] table:t, partition:p1 keep order:false, stats:pseudo",
+	))
+
+	tk1.MustExec(`insert into t values (1), (2), (3), (10), (11)`)
+	tk1.MustExec(`analyze table t with 1 topn, 2 buckets`)
+	tk1.MustQuery("explain format = 'brief' select * from t").Check(testkit.Rows(
+		"TableReader 5.00 root partition:all data:TableFullScan",
+		"└─TableFullScan 5.00 cop[tikv] table:t keep order:false",
+	))
+	tk2.MustQuery("explain format = 'brief' select * from t").Check(testkit.Rows(
+		"PartitionUnion 5.00 root  ",
+		"├─TableReader 3.00 root  data:TableFullScan",
+		"│ └─TableFullScan 3.00 cop[tikv] table:t, partition:p0 keep order:false",
+		"└─TableReader 2.00 root  data:TableFullScan",
+		"  └─TableFullScan 2.00 cop[tikv] table:t, partition:p1 keep order:false",
+	))
```

Comment:
Please then enable prune-mode on tk2 and disable it on tk1 and then check if it works right again.